doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,75 +1,368 @@
+         lvalue-expr
+        | __CPROVER_typed_target(lvalue-expr)
+        | __CPROVER_object_whole(ptr-expr)
+        | __CPROVER_object_from(ptr-expr)
+        | __CPROVER_object_upto(ptr-expr, uint-expr)
 ```
 
-An _assigns_ clause allows the user to specify that a memory location may be written by a function. The set of locations writable by a function is the union of the locations specified by the assigns clauses, or the empty set of no _assigns_ clause is specified. While, in general, an _assigns_ clause could be interpreted with either _writes_ or _modifies_ semantics, this
-design is based on the former. This means that memory not captured by an
-_assigns_ clause must not be written within the given function, even if the
-value(s) therein are not modified.
 
-### Object slice expressions
+For function contracts, the condition and target expressions
+in the assigns clause may only involve function parameters,
+global variables or type identifiers (in `sizeof` or cast expressions).
+The target expression must be free of function calls and side-effects.
+The condition expression may contain calls to side-effect-free functions.
+
+For a loop contract, the condition and target expressions in the assigns clause
+may involve any identifier that is in scope at loop entry
+(parameters of the surrounding function, local variables, global variables,
+type identifiers in `sizeof` or cast expressions, etc.).
+The target expression must be free of function calls and side-effects.
+The condition expression may contain calls to side-effect-free functions.
+
+### Lvalue targets
+
+Roughly speaking, _Lvalues_ are expressions that are associated with a memory location,
+the address of which can be computed using the address-of operator `&`.
+
+Examples of lvalues are: `x` if `x` is either a global or local variable identifier,
+`*ptr` if `ptr` is a pointer expression, `ptr[i]` or `ptr + i` if `ptr` is
+pointer variable or an array and `i` is an integer exression, etc.
+
+Examples of non-lvalues: literal constants like `0`, `1`, ..., arithmetic expressions
+like `x + y` when `x` and `y` are both arithmetic variables,
+function pointer expressions, etc.
 
-The following functions can be used in assigns clause to specify ranges of 
-assignable addresses.
+An lvalue target `expr` with a complete type `expr_t` specifies that the range
+starting at `&expr` and of size `sizeof(expr_t)` bytes is assignable.
+
+Lvalue expressions can also be wrapped in `__CPROVER_typed_target` with the same
+meaning: for an lvalue `expr` with a complete type `expr_t`,
+`__CPROVER_typed_target(expr)` specifies that the range of `sizeof(expr_t)`
+bytes starting at `&expr` is assignable:
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)` 
-specifies that all bytes starting from the given pointer and until the end of 
-the object are assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+void __CPROVER_typed_target(expr_t expr);
 ```
```

Comment:
This explains when one can use that; when does one _have to_ use it? (In other words: why is this syntax even necessary?)

Reply:
This PR was pulled (somehow artificially) from the larger #6887 that introduces using regular C functions as parameterized assigns clause specifications. So this particular one is meant to be used to specify an lvalue target by calling it from a regular C function, but it can also be used in the assigns clause directly.

 If you don't mind I'd like to keep this here and not have to move it back to #6887, everything will make more sense when finally merged

Reply:
Fine with me!